tpu-provisioner/internal/cloud/gke.go

```diff
@@ -560,3 +639,40 @@ func getAnnotation(p *corev1.Pod, key string) string {
 	}
 	return p.Annotations[key]
 }
+
+// nodePoolSelectiveHash attempts to hash information specific to workload requirements.
+// A selective approach is taken to avoid overzealous node pool recreation under circumstances
+// where values might change due to a config or code change in the provisioner.
+// Example scenario where selective hashing is useful:
+// 1. Provisioner is updated to include new upgrade settings.
+// 2. Some node pool goes into a repairing state.
+// 3. The workload Pod goes into an unschedulable state.
+// 4. The code path for ensuring a matching node pool exists is executed.
+func nodePoolSelectiveHash(np *containerv1beta1.NodePool) (string, error) {
+	h := fnv.New32a()
+	var labels map[string]string
+	if np.Config.Labels != nil {
+		labels = make(map[string]string, len(np.Config.Labels))
+		for k, v := range np.Config.Labels {
+			// Exclude the label that tracks the hash.
+			if k == LabelNodePoolHash {
+				continue
+			}
+			labels[k] = v
+		}
+	}
+	npToHash := &containerv1beta1.NodePool{
+		Config: &containerv1beta1.NodeConfig{
+			Spot:        np.Config.Spot,
+			Labels:      labels,
+			Taints:      np.Config.Taints,
```

Comment:
If a node pool is cordoned, we don't necessary want to create a new node pool for the job, as it will violate Bastion logic. 

The problem we had is pods couldn't be scheduled because it has a node-selector that is different from the labels of existing node pools.  `Spot` information is also in the labels. I think we should only use `labels`.

Reply:
I see.